src/miner.cpp

```diff
@@ -351,332 +385,387 @@ CBlockTemplate* CreateNewBlock(const CChainParams& chainparams, const MinerAddre
 
     // If we're given a coinbase tx, it's been precomputed, its fees are zero,
     // so we can't include any mempool transactions; this will be an empty block.
-    bool nonEmptyBlock = !next_cb_mtx;
+    blockFinished = blockFinished && next_cb_mtx;
```

Comment:
Why is this not an `||`? If a `next_cb_mtx` is provided, then the block is finished, but why is this then also conditioned on the current state of the `blockFinished` value? If `next_cb_mtx == std::nullopt` then this will always set `blockFinished` to `false`. The comment here has remained unchanged, but it should at very least be updated to explain what's happening here.

Reply:
Answering this is also blocking.

Reply:
I think this might be a bug in my backport. IIRC I originally preserved `nonEmptyBlock`, and then realised it could be merged with `blockFinished`; the latter however requires a boolean inversion which I think I messed up.

Reply:
Okay, there _is_ a bug here, but it is only observable when `next_cb_mtx` is provided, i.e. the opposite condition to what you give above.

`BlockAssembler::CreateNewBlock` calls `BlockAssembler::resetBlock` which always sets `blockFinished = false`, and then this line is the next mutation of `blockFinished`. So the two happen to be equivalent when no prepared coinbase tx is provided:

| `blockFinished` | `*next_cb_mtx` | `blockFinished && next_cb_mtx` | `blockFinished \|\| next_cb_mtx` |
|---|---|---|---|
| `false` | `false` | `false` | `false` |
| `false` |`true` | `false` | `true` |
| ~~`true`~~ | `false` | `false` | `true` |
| ~~`true`~~ |`true` | `true` | `true` |